File: python/2023/9/history.py
from dataclasses import dataclass

@dataclass
class History:
    values: list[int]

    @classmethod
    def build_from_line(cls, line: str):
        values = [int(value) for value in line.split(' ')]
        return cls(values)
    
    def find_next(self) -> int:
        delta_lines = []
        deltas = self.values
        # Build out the deltas until they are all equal (the step before they're all 0).
        while True:
            deltas = [deltas[i] - deltas[i-1] for i in range(1, len(deltas))]
            delta_lines.append(deltas)
            deltas_are_all_equal = len(set(deltas)) == 1
            if deltas_are_all_equal:
                break
        # Now work backwards to find the next value for each list of deltas.
        while len(delta_lines) > 1:
            deltas = delta_lines.pop()
            # solve such that the next value of the previous deltas is offset from the previous by the last current delta.
            next_value = delta_lines[-1][-1] + deltas[-1]
            delta_lines[-1].append(next_value)
        # Now we have just one list of deltas, and we can apply it to the main values sequence.
        return self.values[-1] + delta_lines[0][-1]

    def __str__(self):
        return ' '.join(str(value) for value in self.values)

File: rust/src/y2023/d9/history.rs
pub struct History {
    values: Vec<i32>,
}

impl History {
    pub fn new_from_line(line: &str) -> Result<History, Box<dyn std::error::Error>> {
        let values = line.split(" ").map(|s| s.parse::<i32>()).collect::<Result<Vec<_>, _>>()?;
        Ok(History {
            values,
        })
    }

    pub fn find_next(&self) -> Result<i32, Box<dyn std::error::Error>> {
        if self.values.is_empty() {
            return Err("History values are empty".into());
        }

        let mut delta_lines: Vec<Vec<i32>> = Vec::new();
        let mut deltas = self.values.clone();

        // Build out the deltas until they are all equal (the step before they're all 0).
        while deltas.len() > 1 {
            deltas = deltas.windows(2).map(|w| w[1] - w[0]).collect();
            delta_lines.push(deltas.clone());

            if deltas.iter().all(|&x| x == deltas[0]) {
                break;
            }
        }

        if delta_lines.is_empty() {
            return Err("No delta lines were generated".into());
        }

        // Now work backwards to find the next value for each list of deltas.
        while delta_lines.len() > 1 {
            let deltas = delta_lines.pop().ok_or("Failed to pop from delta_lines")?;
            let last_value_in_previous = *delta_lines.last_mut()
                .ok_or("No last element in delta_lines")?
                .last().ok_or("No last element in last delta_lines")?;
            let last_delta = *deltas.last().ok_or("No last delta value")?;
            let next_value = last_value_in_previous + last_delta;
            delta_lines.last_mut().unwrap().push(next_value);  // Safe unwrap as we just checked it
        }

        let last_values_delta = delta_lines[0].last().ok_or("No last value in deltas")?;
        Ok(*self.values.last().unwrap() + *last_values_delta)  // Safe unwrap as we checked values is not empty
    }

    pub fn find_previous(&self) -> Result<i32, Box<dyn std::error::Error>> {
        if self.values.is_empty() {
            return Err("History values are empty".into());
        }

        let mut delta_lines: Vec<Vec<i32>> = Vec::new();
        let mut deltas = self.values.clone();

        // Build out the deltas until they are all equal (the step before they're all 0).
        while deltas.len() > 1 {
            deltas = deltas.windows(2).map(|w| w[1] - w[0]).collect();
            delta_lines.push(deltas.clone());

            if deltas.iter().all(|&x| x == deltas[0]) {
                break;
            }
        }

        if delta_lines.is_empty() {
            return Err("No delta lines were generated".into());
        }

        // Now work backwards to find the previous value for each list of deltas.
        while delta_lines.len() > 1 {
            let deltas = delta_lines.pop().ok_or("Failed to pop from delta_lines")?;
            let first_value_in_previous = *delta_lines.last_mut()
                .ok_or("No last element in delta_lines")?
                .first().ok_or("No first element in last delta_lines")?;
            let first_delta = *deltas.first().ok_or("No first delta value")?;
            let prev_value = first_value_in_previous - first_delta;
            delta_lines.last_mut().unwrap().insert(0, prev_value);  // Safe unwrap as we just checked it
        }

        let first_values_delta = delta_lines[0].first().ok_or("No first value in deltas")?;
        Ok(*self.values.first().unwrap() - *first_values_delta)  // Safe unwrap as we checked values is not empty
    }
}